File: android/app/src/main/java/com/cactus/remoteterminal/ui/PairingPayload.kt
package com.cactus.remoteterminal.ui

/**
 * What a scanned pairing QR code is allowed to contain. Kept separate from the
 * scanner screen so the accepted shapes are testable without a camera.
 *
 * Two forms are accepted:
 *  - a bare six-digit code, as the agent prints it and a paired phone shows it;
 *  - a `remoteterminal://pair?relay=…&code=…` link, which also carries the relay.
 *
 * Anything else is rejected: the scanner must not silently take a relay URL
 * from an arbitrary QR code a user happens to point the camera at.
 */
object PairingPayload {

    data class Parsed(val relay: String?, val code: String)

    private val CODE = Regex("^[0-9]{6}$")

    fun parse(text: String?): Parsed? {
        val trimmed = text?.trim().orEmpty()
        if (trimmed.isEmpty()) return null
        if (CODE.matches(trimmed)) return Parsed(null, trimmed)

        val scheme = trimmed.substringBefore("://", missingDelimiterValue = "").lowercase()
        if (scheme != "remoteterminal") return null
        val query = trimmed.substringAfter('?', missingDelimiterValue = "")
        if (query.isEmpty()) return null

        val params = HashMap<String, String>()
        for (pair in query.split('&')) {
            val key = pair.substringBefore('=')
            if (key.isEmpty()) continue
            params[key.lowercase()] = decode(pair.substringAfter('=', missingDelimiterValue = ""))
        }

        val code = params["code"]?.filter { it.isDigit() } ?: return null
        if (!CODE.matches(code)) return null

        val relay = params["relay"]?.takeIf { it.isNotEmpty() }
        // Only relay URLs we would connect to anyway; never an arbitrary scheme.
        if (relay != null && !relay.startsWith("ws://") && !relay.startsWith("wss://")) return null
        return Parsed(relay, code)
    }

    /** Minimal percent-decoding; pairing links carry only a URL and digits. */
    private fun decode(value: String): String {
        if ('%' !in value && '+' !in value) return value
        return try {
            java.net.URLDecoder.decode(value, "UTF-8")
        } catch (_: Exception) {
            value
        }
    }
}

File: desktop/ui/js/core/pairingPayload.js
/**
 * What a pairing payload is allowed to contain. A port of `PairingPayload.kt`,
 * kept separate from the pairing screen so the accepted shapes are testable.
 *
 * Two forms are accepted:
 *  - a bare six-digit code, as the agent prints it and a paired device shows it;
 *  - a `remoteterminal://pair?relay=…&code=…` link, which also carries the relay.
 *
 * Anything else is rejected: pairing must not silently take a relay URL from
 * whatever happens to be on the clipboard.
 */

const CODE = /^[0-9]{6}$/;

export function parsePairingPayload(text) {
  const trimmed = String(text ?? '').trim();
  if (!trimmed) return null;
  if (CODE.test(trimmed)) return { relay: null, code: trimmed };

  const schemeEnd = trimmed.indexOf('://');
  if (schemeEnd < 0) return null;
  if (trimmed.slice(0, schemeEnd).toLowerCase() !== 'remoteterminal') return null;

  const queryStart = trimmed.indexOf('?');
  if (queryStart < 0) return null;
  const query = trimmed.slice(queryStart + 1);
  if (!query) return null;

  const params = new Map();
  for (const pair of query.split('&')) {
    const eq = pair.indexOf('=');
    const key = (eq < 0 ? pair : pair.slice(0, eq)).toLowerCase();
    if (!key) continue;
    params.set(key, decode(eq < 0 ? '' : pair.slice(eq + 1)));
  }

  const code = (params.get('code') ?? '').replace(/\D/g, '');
  if (!CODE.test(code)) return null;

  const relay = params.get('relay') || null;
  // Only relay URLs we would connect to anyway; never an arbitrary scheme.
  if (relay && !relay.startsWith('ws://') && !relay.startsWith('wss://')) return null;
  return { relay, code };
}

/** Minimal percent-decoding; pairing links carry only a URL and digits. */
function decode(value) {
  if (!value.includes('%') && !value.includes('+')) return value;
  try {
    return decodeURIComponent(value.replace(/\+/g, ' '));
  } catch {
    return value;
  }
}
